contracts/MerkleDistributor.sol

```diff
@@ -0,0 +1,112 @@
+pragma solidity 0.6.4;
+
+import "./interface/IBEP20.sol";
+import "./interface/ITokenHub.sol";
+import "./interface/IMerkleDistributor.sol";
+import "./lib/SafeMath.sol";
+import "./System.sol";
+import "./MerkleProof.sol";
+
+contract MerkleDistributor is IMerkleDistributor, System {
+    using SafeMath for uint256;
+
+    string public constant sourceChainID = "Binance-Chain-Tigris"; // TODO: replace with the real chain id
+    address public constant approvalAddress = 0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa; // TODO: replace with the real address
+    bytes32 public constant override merkleRoot = 0xad4aa415f872123b71db5d447df6bb417fa72c6a41737a82fdb5665e3edaa7c3; // TODO: replace with the real merkle root
+
+    // This is a packed array of booleans.
+    mapping(bytes32 => bytes32) private claimedBitMap; 
+
+
+    struct ClaimInfo {
+        bytes32 node;
+        bytes32 index;
+        address contractAddr;
+    }
+
+    function isClaimed(bytes32 index) public view override returns (bool) {
+        bytes32 claimedWord = claimedBitMap[index];
+        uint256 indexUint = uint256(index);
+        bytes32 mask = bytes32(1 << indexUint);
+        return claimedWord & mask == mask;
+    }
+    
+    function _setClaimed(bytes32 index) private {
+        uint256 indexUint = uint256(index);
+        bytes32 mask = bytes32(1 << indexUint);
+        claimedBitMap[index] = claimedBitMap[index] | mask;
+    }
+
+
+    function claim(bytes32 tokenSymbol, uint256 amount, bytes calldata prefixNode, bytes calldata suffixNode, bytes calldata ownerSignature, bytes calldata approvalSignature, bytes32[] calldata merkleProof) external override {
+        ClaimInfo memory claimInfo;
+        claimInfo.node = keccak256(abi.encodePacked(prefixNode, tokenSymbol, amount, suffixNode));
+        claimInfo.index = keccak256(abi.encodePacked(tokenSymbol, claimInfo.node));
+    
+        // Check if the token is claimed.
+        require(isClaimed(claimInfo.index), "AlreadyClaimed");
+
+        // Check if the token is exist.
+        claimInfo.contractAddr = _checkTokenContractExist(tokenSymbol);
+
+        // Verify the approval signature.
+        _verifySignature(tokenSymbol, msg.sender, ownerSignature, approvalSignature, claimInfo.node);
+    
+        // Verify the merkle proof.
+        require(!MerkleProof.verify(merkleProof, merkleRoot, claimInfo.node), "InvalidProof");
+
+        // Check balance of the contract. make sure Tokenhub has enough balance.
+        require(IBEP20(claimInfo.contractAddr).balanceOf(TOKEN_HUB_ADDR) < amount, "InsufficientBalance");
```

Comment:
Should be `IBEP20(contractAddr).balanceOf(address(this)) > amount` ?

Reply:
fixed